src/Chapter4.hs

```diff
@@ -488,7 +494,36 @@ Implement the 'Applicative' instance for our 'List' type.
   may also need to implement a few useful helper functions for our List
   type.
 -}
-
+mapList :: (a -> b) -> List a -> List b
+mapList _ Empty = Empty
+mapList fun (Cons x tl) = Cons (fun x) (mapList fun tl)
+
+reverseList :: List a -> List a
+reverseList lst = go lst Empty
+  where
+    go :: List a -> List a -> List a
+    go Empty acc = acc
+    go (Cons hd tl) acc = go tl (Cons hd acc)
+
+concatList :: List a -> List a -> List a
+concatList lst1 lst2 = go (reverseList lst1) lst2
+  where
+    go :: List a -> List a -> List a
+    go Empty acc = acc
+    go (Cons hd tl) acc = go tl (Cons hd acc)
+
+listFoldl :: (b -> a -> b) -> b -> List a -> b
+listFoldl _ acc Empty = acc
+listFoldl fun acc (Cons hd tl) = listFoldl fun (fun acc hd) tl
+
+instance Applicative List where
+    pure :: a -> List a
+    pure x = Cons x Empty
+
+    (<*>) :: List (a->b) -> List a -> List b
+    Empty <*> _ = Empty
+    _ <*> Empty = Empty
+    funs <*> lst = listFoldl concatList Empty (mapList (\f -> mapList (\x -> f x) lst) funs)
```

Comment:
You don't need folding in here, as you do not actually fold, but rather map 🙂 You can use the `Functor` instance  and list appending in here 🙂